src/power.jl

```diff
@@ -58,37 +58,95 @@ end
 
 ^(a::Taylor1{TaylorN{T}}, r::Rational) where {T<:NumberNotSeries} = a^(r.num/r.den)
 
+# in-place form of power_by_squaring
+# this method assumes `y`, `x` and `aux1` are of same order
+function power_by_squaring!(y::TaylorN{T}, x::TaylorN{T}, aux1::TaylorN{T}, p::Integer) where {T<:NumberNotSeries}
+    t = trailing_zeros(p) + 1
+    p >>= t
+    # aux1 = x
+    for k in eachindex(aux1)
+        identity!(aux1, x, k)
+    end
+    while (t -= 1) > 0
+        # aux1 = square(aux1)
+        for k in reverse(eachindex(aux1))
+            sqr!(aux1, k)
+        end
+    end
+    # y = aux1
+    for k in eachindex(y)
+        identity!(y, aux1, k)
+    end
+    while p > 0
+        t = trailing_zeros(p) + 1
+        p >>= t
+        while (t -= 1) ≥ 0
+            # aux1 = square(aux1)
+            for k in reverse(eachindex(aux1))
+                sqr!(aux1, k)
+            end
+        end
+        # y = y * aux1
+        mul!(y, aux1)
+    end
+    return nothing
+end
 
 
 # power_by_squaring; slightly modified from base/intfuncs.jl
 # Licensed under MIT "Expat"
 for T in (:Taylor1, :HomogeneousPolynomial, :TaylorN)
-    @eval function power_by_squaring(x::$T, p::Integer)
-        if p == 1
-            return copy(x)
-        elseif p == 0
-            return one(x)
-        elseif p == 2
-            return square(x)
-        end
-        t = trailing_zeros(p) + 1
-        p >>= t
-        while (t -= 1) > 0
-            x = square(x)
-        end
-        y = x
-        while p > 0
+    @eval power_by_squaring(x::$T, p::Integer) = power_by_squaring(x, Val(p))
+    @eval power_by_squaring(x::$T, ::Val{0}) = one(x)
+    @eval power_by_squaring(x::$T, ::Val{1}) = copy(x)
+    @eval power_by_squaring(x::$T, ::Val{2}) = square(x)
```

Comment:
Perhaps adding `power_by_squaring(x::$T, ::Val{3})` could be useful too, since it includes the "odd" powers part of the function.

Reply:
You mean adding something along the lines of `power_by_squaring(x::$T, ::Val{3}) = x*square(x)`?

Reply:
Good point... I guess this is the kind of question that we should define in terms of performance. My naive guess is that `x*square(x)` will allocate more than the direct call to `power_by_squaring`, but we should definetily check this.

Reply:
I just checked this, and it turns out doing `x*square(x)` allocates less than `power_by_squaring(x)` (!), do you think it's worth then to add the `::Val{3}` method?

Reply:
Have you also teste with intervals? I would choose the case that behaves better wrt allocations.

Reply:
With intervals `power_by_squaring(x, 3)` on `Taylor1{Interval{Float64}}` allocates around half as `x*square(x)`, while for `TaylorN{Interval{Float64}}` allocations are essentially the same